plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClientModule.java

```diff
@@ -48,8 +67,53 @@ public void configure(Binder binder)
     @Provides
     @Singleton
     @ForBaseJdbc
-    public ConnectionFactory getConnectionFactory(BaseJdbcConfig config, CredentialProvider credentialProvider)
+    public ConnectionFactory getConnectionFactory(BaseJdbcConfig config, SqlServerConfig sqlServerConfig, CredentialProvider credentialProvider)
+    {
+        return new ConfiguringConnectionFactory(
+                new DriverConnectionFactory(new SQLServerDriver(), config, credentialProvider),
+                connection -> configureConnectionTransactionIsolation(sqlServerConfig.isSnapshotIsolationDisabled(), connection));
+    }
+
+    public void configureConnectionTransactionIsolation(boolean snapshotIsolationDisabled, Connection connection)
+            throws SQLException
+    {
+        if (snapshotIsolationDisabled) {
+            return;
+        }
+        try {
+            if (hasSnapshotIsolationEnabled(connection)) {
+                // SQL Server's READ COMMITTED + SNAPSHOT ISOLATION is equivalent to ordinary READ COMMITTED in e.g. Oracle, PostgreSQL.
+                connection.setTransactionIsolation(TRANSACTION_SNAPSHOT);
+            }
+        }
+        catch (SQLException e) {
+            connection.close();
+            throw e;
+        }
+    }
+
+    private boolean hasSnapshotIsolationEnabled(Connection connection)
+            throws SQLException
+    {
+        try {
+            return snapshotIsolationEnabled.get(SnapshotIsolationEnabledCacheKey.INSTANCE, () -> {
+                Handle handle = Jdbi.open(connection);
```

Comment:
Does this work properly with things like impersonation? i.e. if extra credentials are provided then is it possible that user doesn't have permissions to select from sys.databases?

Should there be a kill-switch?

Reply:
Snapshot isolation is already controlled by `sqlserver.snapshot-isolation.disabled` config property?